### PR TITLE
feat: Table toolbar v10

### PR DESCRIPTION
**Pull request by vpicone**
Closes IBM/carbon-components-react#2166

Add support for new v10 features

TODO:
- [x] Update markup
- [x] Update classnames
- [x] Create menu component
- [x] Update action component
- [x] TableToolbarSearch expansion needs should be a controlled/uncontrolled component
- [x] Search filtering
- [x] Batch Actions
- [x] Update stories for new markup
- [x] Switch toolbar for menu in all stories
- [x] add new table modifiers and remove old ones
- [x] add knobs to stories
- [x] focus search input when expanded
- [x] adjust default story and prop naming

#### Changelog

**New**

- TableToolbarMenu
- Table modifiers
- Story knobs

**Changed**

- Actions are now overflow menu items that live within a single menu
- Slight changes in markup requiring data table 

**Removed**

- Deprecation warnings

### CODE REVIEW

**Comment by vpicone** [comment]
@joshblack I'm having issues getting the search filtering to work. I'm having trouble figuring out how the test/filtering process is meant to go down.

File: src/components/DataTable/TableToolbarMenu.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  );
+};
+
+TableToolbarMenu.defaultProps = {
+  renderIcon: Settings16,
+  iconDescription: 'Settings',
+};
+
+TableToolbarMenu.propTypes = {
+  children: PropTypes.node.isRequired,
```
Comment (asudoh) [inline src/components/DataTable/TableToolbarMenu.js]: Please add prop description comments like other components do.

File: src/components/DataTable/TableToolbarSearch.js [hunk 2]
```diff
@@ -27,24 +27,63 @@ const translateWithId = id => {
...
+    <div
+      onClick={onClick}
+      onFocus={event => handleExpand(event, true)}
+      onBlur={event => !value && handleExpand(event, false)}
+      className={searchContainerClasses}>
       <Search
-        light
-        className={className}
+        ref={searchRef}
+        className={className + ' ' + `${prefix}--search-maginfier`}
```
Comment (asudoh) [inline src/components/DataTable/TableToolbarSearch.js]: Looks that we should use `classnames` library here.

**Comment by vpicone** [comment]
@asudoh fixed thank you!

**Comment by asudoh** [comment]
Thanks @vpicone for your quick update! Would you update the snapshot?

**Comment by vpicone** [comment]
@dakahn The TableToolBarSearch is a weird component (button-div encapsulating/hiding an input). I'm happy with the keyboard navigation but definitely open to an a11y insight. I played with having the div be a button, but styling get really wild with an input nested in a button. As a result, I needed to add a tabindex="0" to make the div tabable and direct focus to the input.

**Comment by dakahn** [comment]
@vpicone Thanks for the ping! 🏄🏽‍♂️

Testing in latest Chrome, Firefox, on macOS, and Windows 10 on IE11, latest Chrome, Firefox there are five points of interest:
- [ ] inadequate/inconsistent focus state on dropdown menu button on click
- [ ] when dropdown is activated via keyboard the first menuItem should receive focus
Apr-22-2019 09-44-22
- [ ] there is no grouping element telling a user they're currently interacting with a landmarked group _data table toolbar_. Making skipping to table content/returning to toolbar frustrating
Apr-22-2019 09-47-59
- [ ] when the user inputs a filter/search there is no feedback indicating the table has been updated as a live region
Apr-22-2019 09-47-07
- [ ] Funny lookin' business on IE11. Resizing DataTable and cutoff search bar focus style
Apr-22-2019 09-54-32

**Comment by vpicone** [comment]
> @vpicone Thanks for the ping! 🏄🏽‍♂️
> 
> Testing in latest Chrome on macOS, and Windows 10 on IE11/latest Chrome are five points of interest:
> 
>     * [ ]  inadequate/inconsistent focus state on dropdown menu button on click
> 
>     * [ ]  when dropdown is activated via keyboard the first menuItem should receive focus
>       Apr-22-2019 09-44-22
> 
>     * [x]  there is no grouping element telling a user they're currently interacting with a landmarked group _data table toolbar_. Making skipping to table content/returning to toolbar frustrating (in the example the unordered list is our sidebar)
>       Apr-22-2019 09-47-59
> 
>     * [x]  when the user inputs a filter/search there is no feedback indicating the table has been updated as a live region
>       Apr-22-2019 09-47-07
> 
>     * [ ]  Funny lookin' business on IE11. Resizing DataTable and cutoff search bar focus style
>       Apr-22-2019 09-54-32


@dakahn I'm going to make a new issue for the first two incomplete tasks since I believe they're inherent to the overflow menu itself and might require breaking changes.

If you don't mind making an issue for the ie11 one, I don't have ie11 going atm so can't reproduce or debug

**Comment by jnm2377** [comment]
It's looking really good! I noticed the IE11 bug that @dakahn was talking about is also happening in Chrome when you type in the search bar. 
toolbar

**Comment by vpicone** [comment]
@jnm2377 the size of the table is based off of the wrapper width and the contents. Since we don't define a wrapper here, the table is only stretched by the width of its rows. When there's no content, there's no width (up until the min-width)

Search isn't working in vanilla, but if you delete the tr you can see the same thing happening here: https://www.carbondesignsystem.com/components/data-table/code

File: src/components/DataTable/TableToolbar.js [hunk 3]
```diff
@@ -4,16 +4,14 @@
...
 const { prefix } = settings;
 
-const TableToolbar = wrapComponent({
-  name: 'TableToolbar',
-  type: 'section',
-  className: `${prefix}--table-toolbar`,
-});
-
+const TableToolbar = ({ children }) => (
+  <section role="group" className={`${prefix}--table-toolbar`}>
```
Comment (dakahn) [inline src/components/DataTable/TableToolbar.js]: @vpicone I can open those tickets no problem! 

We're close! `<section>` just needs two more attributes to make it sing👩🏽‍🎤: 
```suggestion
  <section role="group" aria-label="data table toolbar" tabIndex="0" className={`${prefix}--table-toolbar`}>
```

With the above -- our grouping of tools is keyboard navigable and will sound like _"main, data table toolbar, group. you are currently on a group --"_ with VoiceOver and variations on that theme for NVDA, JAWS

Comment (dakahn) [reply src/components/DataTable/TableToolbar.js]: I'm not sure if this is a bugged I missed yesterday morning, or if this is newly introduced via recent changes, but the overflow menu button has _even stranger_ behavior:

Apr-23-2019 00-55-56

When the menu button has focus the user is instructed to move through the menu with the left and right arrow keys -- at which point they find the only other keyboard navigable item is the cog svg 😕

Comment (vpicone) [reply src/components/DataTable/TableToolbar.js]: Yeah it's just a wrapper for the OverflowMenu with a different default icon. There's a _lot_ of open issues around that component I think it should take priority for the next time someone has some extra time.

**Comment by vpicone** [comment]
LORD SAVE ME FROM THESE SNAPSHOTS

**Pull request merged**